xarray/core/dataset.py

```diff
@@ -1433,11 +1455,16 @@ def __setitem__(self, key: Hashable, value) -> None:
         variable.
         """
         if utils.is_dict_like(key):
-            raise NotImplementedError(
-                "cannot yet use a dictionary as a key to set Dataset values"
-            )
-
-        self.update({key: value})
+            # loop over dataset variables
+            for var, da in self.items():
+                val = value
+                if type(value) == xr.core.dataset.Dataset:
+                    val = value[var]
+                # only set value if all dimensions are present
+                if all([k in da.dims for k in key.keys()]):
+                    da[key] = val
```

Comment:
I have a couple of concerns with the current implementation:

1. Skipping variables that do not possess all dimensions seems potentially error prone -- I wonder if we should raise an error instead?
2. We should also definitely check that all keys in `value` are used, which could also hide errors.
3. Indexing assignment like `da[key] = val` is not guaranteed to succeed, e.g., if there are mis-matched dimensions an error is raised. It would be better to do all those checks up front, so `ds[key] = value` always either succeeds or raises an error, rather than leaving the Dataset in some in partially updated state.

Reply:
Good point on (3).

While (2) is an easy choice, on (1) I was thinking we'd mirror `__getitem__`, which skips them (indeed, @matzegoebel, could we share some of the implementation with the code in `__getitem__`)?

```python

In [4]: ds
Out[4]:
<xarray.Dataset>
Dimensions:  (lat: 25, lon: 53, time: 2920)
Coordinates:
  * lat      (lat) float32 75.0 72.5 70.0 67.5 65.0 ... 25.0 22.5 20.0 17.5 15.0
  * lon      (lon) float32 200.0 202.5 205.0 207.5 ... 322.5 325.0 327.5 330.0
  * time     (time) datetime64[ns] 2013-01-01 ... 2014-12-31T18:00:00
Data variables:
    air      (time, lat, lon) float32 241.2 242.5 243.5 ... 296.5 296.2 295.7
    air2     (time, lon) float32 279.4 279.7 279.7 279.7 ... 279.4 280.0 280.5
Attributes:
    Conventions:  COARDS
    title:        4x daily NMC reanalysis (1948)
    description:  Data is from NMC initialized reanalysis\n(4x/day).  These a...
    platform:     Model
    references:   http://www.esrl.noaa.gov/psd/data/gridded/data.ncep.reanaly...

In [5]: ds.loc[dict(lat=75)]
Out[5]:
<xarray.Dataset>
Dimensions:  (lon: 53, time: 2920)
Coordinates:
    lat      float32 75.0
  * lon      (lon) float32 200.0 202.5 205.0 207.5 ... 322.5 325.0 327.5 330.0
  * time     (time) datetime64[ns] 2013-01-01 ... 2014-12-31T18:00:00
Data variables:
    air      (time, lon) float32 241.2 242.5 243.5 244.0 ... 241.7 241.5 241.8
    air2     (time, lon) float32 279.4 279.7 279.7 279.7 ... 279.4 280.0 280.5
Attributes:
    Conventions:  COARDS
    title:        4x daily NMC reanalysis (1948)
    description:  Data is from NMC initialized reanalysis\n(4x/day).  These a...
    platform:     Model
    references:   http://www.esrl.noaa.gov/psd/data/gridded/data.ncep.reanaly...
```

Reply:
Thanks for the example. Concretely: what should `ds.loc[dict(lat=75)] = 0.0` mean in this scenario for the variable `air2`?

Arguably, the variable `air2` should be entirely set to zero rather than skipped. The normal meaning of `x[y] = z` is that `x[y]` should be set equal to `z`, such that after setting an item, getting an item should return the set value.

@max-sixty  I think this is what you meant by your comment, but it isn't what is implemented in the current code, which would not update the value of `air2` at all.

Reply:
V good point. Thanks.

Either a) raising on missing dimensions or b) "entirely set to zero" in this case, would be reasonable imo.

To the extent you're less confident that (b) is correct, I'd suggest we move forward with (a) and evaluating whether we should do (b) separately.

(though @shoyer if you're up for it, let's discuss the general pace of reviews next week and whether you think this sort of "narrowing" of PR scope is a reasonable tactic)

Reply:
In my opinion, setting air2 entirely to the new data doesn't make sense in most cases. It's certainly not what I would expect if the variable has no relation to latitude at all. And I might not realize the change immediately if the new data also depends on longitude and time, so that's potentially dangerous. I agree that it's best raising an error as the default option. But the error message could hint at an xarray setting that allows ignoring such cases and leaving those variables untouched.

Reply:
But that optional xarray setting should be a separate PR, I guess

Reply:
Perfect, let's go ahead with the limited solution for now.

I agree it would be a bit surprising to set the whole variable. But I'm not sure skipping over it is coherent with the principle re "The normal meaning of x[y] = z is that x[y] should be set equal to z, such that after setting an item, getting an item should return the set value."...